File: Src/java/engine/src/test/java/org/opencds/cqf/cql/engine/execution/CqlExternalFunctionsTest.java
package org.opencds.cqf.cql.engine.execution;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;

import java.util.Arrays;
import org.hl7.elm.r1.VersionedIdentifier;
import org.opencds.cqf.cql.engine.data.SystemExternalFunctionProvider;
import org.opencds.cqf.cql.engine.execution.external.MyMath;
import org.testng.annotations.Test;

public class CqlExternalFunctionsTest extends CqlTestBase {

    @Test
    public void testExternalFunctions() {
        VersionedIdentifier identifier = toElmIdentifier("CqlExternalFunctionsTest");

        engine.getState()
                .getEnvironment()
                .registerExternalFunctionProvider(
                        identifier,
                        new SystemExternalFunctionProvider(Arrays.asList(MyMath.class.getDeclaredMethods())));

        var results = engine.evaluate(identifier);
        var value = results.forExpression("CallMyPlus").value();
        assertThat(value, is(10));

        value = results.forExpression("CallMyMinus").value();
        assertThat(value, is(-2));
    }
}

File: Src/java/engine/src/test/kotlin/org/opencds/cqf/cql/engine/execution/CqlExternalFunctionsTest.kt
package org.opencds.cqf.cql.engine.execution

import java.lang.reflect.Method
import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers
import org.junit.jupiter.api.Test
import org.opencds.cqf.cql.engine.data.SystemExternalFunctionProvider
import org.opencds.cqf.cql.engine.execution.external.MyMath

internal class CqlExternalFunctionsTest : CqlTestBase() {
    @Test
    fun externalFunctions() {
        val identifier = toElmIdentifier("CqlExternalFunctionsTest")

        val engine = engine
        engine.state.environment.registerExternalFunctionProvider(
            identifier,
            SystemExternalFunctionProvider(
                listOf<Method?>(*MyMath::class.java.getDeclaredMethods())
            ),
        )

        val results = engine.evaluate(identifier)
        var value = results.forExpression("CallMyPlus").value()
        MatcherAssert.assertThat(value, Matchers.`is`(10))

        value = results.forExpression("CallMyMinus").value()
        MatcherAssert.assertThat(value, Matchers.`is`(-2))
    }
}
